### PR TITLE
Use PC instead of scalariform for `ExpandSelectionReq`

### LINKED ISSUE #1622: Rewrite expand region without scalariform
With #1619  done, the only functionality using scalariform is expand region. Rewriting that would enable us to completely remove scalariform from the server and place it in the build tool plugins.

### PR DESCRIPTION
**Pull request by sugakandrey**
close #1622 
The most straightforward implementation of `ExpandSelection` feature using pres. compiler `Locator` modified to return smallest tree with a position **strictly** bigger than given. The downside is that we traverse the all the trees every time client asks to expand/contract. Perhaps we can just collect all the trees, which include the initial position, and make client use them for repeating expand/contract commands. 

Also, this will require some `ensime-emacs` adjustments, current implementation copies file contents into a temporary file, which is then loaded into pres. compiler along with the real one, leading to `duplicate definition`s all over the code.

### CODE REVIEW

**Comment by fommil** [comment]
Hmm, W not too keen on an api change for this. Can we unload and reload the file or do something more like scala-ide where it just parses the code bust doesn't typecheck, like @dragos recommended.

**Comment by fommil** [comment]
Hehe, nice. No api change is good though. Parsing the fire is basically free.

**Comment by sugakandrey** [comment]
alright, should be better now :)

File: core/src/it/scala/org/ensime/intg/BasicWorkflow.scala [hunk 1]
```diff
@@ -245,13 +245,13 @@ class BasicWorkflow extends EnsimeSpec
           )
 
           // expand selection around 'val foo'
-          project ! ExpandSelectionReq(fooFile, 215, 215)
+          project ! ExpandSelectionReq(fooFile, 290, 290)
```
Comment (fommil) [inline core/src/it/scala/org/ensime/intg/BasicWorkflow.scala]: why did you need to change the expectations?

File: project/EnsimeBuild.scala [hunk 2]
```diff
@@ -71,9 +71,6 @@ object EnsimeBuild {
   )
 
   lazy val api = Project("api", file("api")) settings (commonSettings) settings (
-    libraryDependencies ++= Seq(
-      "org.scalariform" %% "scalariform" % "0.1.8"
```
Comment (fommil) [inline project/EnsimeBuild.scala]: :tada:

File: core/src/main/scala/org/ensime/core/PositionLocator.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// Copyright: 2010 - 2016 https://github.com/ensime/ensime-server/graphs
+// License: http://www.gnu.org/licenses/gpl-3.0.en.html
+package org.ensime.core
+
+import scala.tools.nsc.interactive.Global
+
+trait PositionLocator {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/core/PositionLocator.scala]: if this file has been derived from scala-ide it needs to be properly attributed to them

File: core/src/it/scala/org/ensime/intg/BasicWorkflow.scala (hunk 1 above)
Comment (sugakandrey) [inline core/src/it/scala/org/ensime/intg/BasicWorkflow.scala]: PC implementation of `expand-selection` is not 100% equivalent to scalariform.
For example `val f\*cursor here*\oo = new Foo()` scalariform would expand to just the name `foo`, while the smallest tree is the whole `ValDef`.

Comment (dragos) [inline core/src/it/scala/org/ensime/intg/BasicWorkflow.scala]: Yeah, that's a bummer. I think scala-refactoring has some code to reliably find positions of names (for instance, for Rename). Otherwise could also handle this "on site" (same goes for all definitions, I imagine).

**Comment by fommil** [comment]
still flagged as WIP, let me know when you're happy with it... I'm not too keen on changing the test expectation. Adding to it is better.

**Comment by sugakandrey** [comment]
I'm looking into scala-refactoring as @dragos suggested, pretty sure they have what we need.

**Comment by sugakandrey** [comment]
@fommil Looks fine to me now.

**Comment by fommil** [comment]
Awesome!

**Comment by fommil** [comment]
this feels snappier, good stuff!

**Comment by fommil** [comment]
I'm tempted to increase the API here so that editors can do more, e.g. a way to ask for all blocks in a file... would allow structural editing. Like in Emacs we could have something like smart-parens but specifically for implied scala brackets.

**Comment by fommil** [comment]
protip, combine with

```elisp
(defun scala-sp-blockify-region ()
  (interactive)
  (when (region-active-p)
    (kill-region (region-beginning) (region-end))
    (insert ?{) (insert ?}) (left-char)
    (comment-indent-new-line)
    (comment-indent-new-line)
    (forward-line -1)
    (yank)
    (call-interactively 'indent-region)))
(bind-key "C-+" 'scala-sp-blockify-region)
```

for awesome `+++ C-+` workflow to blockify / indent regions.
